File: bindings/dotnet/src/Zlink/Contracts/Sockets/PubSubSocketContracts.cs
// SPDX-License-Identifier: MPL-2.0

using System;

namespace Systems.Zlink;

/// <summary>
/// Defines the publisher socket contract.
/// </summary>
public interface IPublisherSocket : IConnectableSocket
{
    /// <summary>
    /// Starts a publish operation.
    /// </summary>
    SendOperation Publish(string topic);

    /// <summary>
    /// Registers a handler for send ready.
    /// </summary>
    void OnSendReady(Action handler);
}

/// <summary>
/// Defines the subscriber socket contract.
/// </summary>
public interface ISubscriberSocket : IConnectableSocket
{
    /// <summary>
    /// Sets the subscription.
    /// </summary>
    void SetSubscription(string topicOrPattern);

    /// <summary>
    /// Unsets the subscription.
    /// </summary>
    void UnsetSubscription(string topicOrPattern);

    /// <summary>
    /// Gets the subscription at the specified index.
    /// </summary>
    SubscriptionEntry? SubscriptionAt(int index);

    /// <summary>
    /// Receives a subscribed topic message.
    /// </summary>
    bool Subscribe(TopicMessage result, RecvFlags flags = RecvFlags.None);
}

/// <summary>
/// Defines the pub socket contract.
/// </summary>
public interface IPubSocket : IPublisherSocket
{
    /// <summary>
    /// Gets the options.
    /// </summary>
    new PubSocketOptions Options { get; }

    /// <summary>
    /// Attaches a discovery service.
    /// </summary>
    void AttachDiscovery(IDiscovery discovery);
}

/// <summary>
/// Defines the sub socket contract.
/// </summary>
public interface ISubSocket : ISubscriberSocket
{
    /// <summary>
    /// Gets the options.
    /// </summary>
    new SubSocketOptions Options { get; }

    /// <summary>
    /// Attaches a discovery service.
    /// </summary>
    void AttachDiscovery(IDiscovery discovery);
}

/// <summary>
/// Defines the xpub socket contract.
/// </summary>
public interface IXPubSocket : IPublisherSocket
{
    /// <summary>
    /// Gets the options.
    /// </summary>
    new PubSocketOptions Options { get; }

    /// <summary>
    /// Receives the next available item.
    /// </summary>
    bool ReceiveSubscriptionEvent(SubscriptionEvent result,
        RecvFlags flags = RecvFlags.None);
}

/// <summary>
/// Defines the xsub socket contract.
/// </summary>
public interface IXSubSocket : ISubscriberSocket
{
    /// <summary>
    /// Gets the options.
    /// </summary>
    new SubSocketOptions Options { get; }
}

File: bindings/python/src/zlink/contracts/sockets/pubsub_socket_contracts.py
# SPDX-License-Identifier: MPL-2.0

from . import socket as _socket_contract


class PubSocket(_socket_contract._SocketContract):
    def __new__(cls, context=None):
        if cls is PubSocket:
            return _socket_contract._require(
                _socket_contract._pub_socket_factory, "pub socket"
            )(context)
        return object.__new__(cls)

    @property
    def pub_options(self): ...

    def publish(self, topic): ...


class SubSocket(_socket_contract._SocketContract):
    def __new__(cls, context=None):
        if cls is SubSocket:
            return _socket_contract._require(
                _socket_contract._sub_socket_factory, "sub socket"
            )(context)
        return object.__new__(cls)

    @property
    def sub_options(self): ...

    def subscribe(self, topic): ...

    def unsubscribe(self, topic): ...

    def recv(self, *, flags=0): ...

    def recv_into(self, topic_message, *, flags=0): ...


class XPubSocket(_socket_contract._SocketContract):
    def __new__(cls, context=None):
        if cls is XPubSocket:
            return _socket_contract._require(
                _socket_contract._xpub_socket_factory, "xpub socket"
            )(context)
        return object.__new__(cls)

    @property
    def pub_options(self): ...

    def publish(self, topic): ...

    def receive_subscription_event_into(self, event, *, flags=0): ...


class XSubSocket(_socket_contract._SocketContract):
    def __new__(cls, context=None):
        if cls is XSubSocket:
            return _socket_contract._require(
                _socket_contract._xsub_socket_factory, "xsub socket"
            )(context)
        return object.__new__(cls)

    @property
    def sub_options(self): ...

    def subscribe(self, topic): ...

    def unsubscribe(self, topic): ...


__all__ = ["PubSocket", "SubSocket", "XPubSocket", "XSubSocket"]
